ably/realtime_channel.go

```diff
@@ -2,130 +2,342 @@ package ably
+// capacity and pass it as ch; the Subscribe method will return it instead of
+// creating new one.
+//
+// If ch is non-nil and it was already registered to receive messages with different
+// names than the ones given, it will be added to receive also the new ones.
+func (c *RealtimeChannel) Subscribe(ch chan *proto.Message, names ...string) (chan *proto.Message, error) {
+	if err := c.Attach(); err != nil {
+		return nil, err
+	}
+	if ch == nil {
+		ch = make(chan *proto.Message)
+	}
+	if len(names) == 0 {
+		names = []string{""}
+	}
 	c.listenMtx.Lock()
-	if _, ok := c.listeners[event]; !ok {
-		c.listeners[event] = make(map[chan *proto.Message]struct{})
+	for _, name := range names {
+		l, ok := c.listen[name]
+		if !ok {
+			l = make(map[chan *proto.Message]struct{})
+			c.listen[name] = l
+		}
+		l[ch] = struct{}{}
 	}
-	c.listeners[event][ch] = struct{}{}
 	c.listenMtx.Unlock()
-	go c.Attach()
-	return ch
+	return ch, nil
 }
 
-func (c *RealtimeChannel) Unsubscribe(event string, ch chan *proto.Message) {
+// Unsubscribe removes previously registered channel for the given message names.
+func (c *RealtimeChannel) Unsubscribe(ch chan *proto.Message, names ...string) {
+	if ch == nil {
+		panic("ably: RealtimeChannel.Unsubscribe using nil channel")
+	}
+	if len(names) == 0 {
+		names = []string{""}
+	}
 	c.listenMtx.Lock()
-	delete(c.listeners[event], ch)
-	if len(c.listeners[event]) == 0 {
-		delete(c.listeners, event)
+	for _, name := range names {
+		delete(c.listen[name], ch)
+		if len(c.listen[name]) == 0 {
+			delete(c.listen, name)
+		}
 	}
 	c.listenMtx.Unlock()
-	close(ch)
 }
 
-func (c *RealtimeChannel) Publish(name string, data string) error {
-	c.Attach()
+// On relays request channel states to c; on state transition
+// connection will not block sending to c - the caller must ensure the incoming
+// values are read at proper pace or the c is sufficiently buffered.
+//
+// If no states are given, c is registered for all of them.
+// If c is nil, the method panics.
+// If c is alreadt registered, its state set is expanded.
+func (c *RealtimeChannel) On(ch chan<- State, states ...int) {
+	c.state.on(ch, states...)
+}
+
+// Off removes c from listetning on the given channel state transitions.
+//
+// If no states are given, c is removed for all of the connection's states.
+// If c is nil, the method panics.
+func (c *RealtimeChannel) Off(ch chan<- State, states ...int) {
+	c.state.off(ch, states...)
+}
+
+// Publish publishes a message on the channel, which is send on separate
+// goroutine. Publish does not block.
+//
+// If listen channel is non-nil it will be used to notify whether the message
+// was published succesfully with nil error value; when the operation failed,
+// the error that caused the failure is going to be sent on the listen channel.
+//
+// This implicitly attaches the channel if it's not already attached.
+func (c *RealtimeChannel) Publish(name string, data string, listen chan<- error) error {
```

Comment:
The third argument here seems quite awkward if I want to publish a message and wait for it:

```go
errCh := make(chan error)
if err := c.Publish("foo", "bar", errCh); err != nil {
  return err
}
if err := <-errCh; err != nil {
  return err
}
```

I think it would be clearer with something like:

```go
res, err := c.Publish("foo", "bar")
if err != nil {
  return err
}
res.Wait()
if res.Failed {
  return res.Err
}